### DISCUSSION TITLE
Multiple saved IPs and swapping between them

### DISCUSSION THREAD

**Discussion by crazy-hair** [suchmememanyskill/CYD-Klipper#41] [Ideas]
I have a lot of printers, and buying a separate display for each would be quite costly. Having the option to save a bunch of IPs and then switch between them would be super useful, as I'm usually only fiddling with one printer at a time (maybe having different UI colors for different profiles to prevent confusion?) Anyway, don't know how much of a hassle it would be to implement, but just want to know what you think. Thanks!

**Comment by suchmememanyskill** [comment]
I don't think it would be too difficult to create different configuration profiles (as in, just store multiple settings and switch between them). You will probably quickly hit storage limits though

Although i don't think you can create a nice user experience around this; constantly switching printers is gonna suck on such a small screen

**Reply by crazy-hair** [reply]
I would think maybe having a button per printer under the macros list would make it simple enough to pop between them (essentially the same way the macros are set up right now, except a list of printers instead of a list of macros)

**Reply by jan610** [reply]
I second this !
I love this thing and made my CYD battery powered and would like to monitor my two printers while watching TV or being on the Balcony for example, but don't want to buy/build/bring two of those devices with me.
And using my mobilephone is out of question ofc ^^

**Reply by suchmememanyskill** [reply]
Alright, i'll give it a shot.
Maybe it's worth making an overview screen where you can see the state and % done of all printers

**Comment by suchmememanyskill** [comment]
@CrazyHairGuy @jan610 Proposed configuration per printer:
- IP/Port/Auth
- Light mode & Invert colors
- Remaining time calc mode
- Printer name
- Color scheme
- Hotend/Bed temp presets

Configuration global:
- Screen calibration
- Wifi config
- Screen rotation
- Auto OTA update
- Multi printer mode (proposed: will modify the UI so multiple printers can be controlled
- On during print
- Brightness
- Screen timeout

**Reply by suchmememanyskill** [reply]
PS: This does sadly mean i have to wipe everyone's configs again next update, the data structure changed too much this update to be backwards compatible

**Reply by suchmememanyskill** [reply]
![PXL_20240307_231322520_1.jpg](https://github.com/suchmememanyskill/CYD-Klipper/assets/38142618/7692bf8d-8d2a-4396-a197-94ca7520f194)

**Reply by crazy-hair** [reply]
This is perfect! Thank you for putting so much effort into this project.

**Reply by suchmememanyskill** [reply]
@crazy-hair @jan610 Would you mind testing the following build and giving me some feedback (in how it works, things you find confusing, etc):

Actions build (use to download merged firmware): https://github.com/suchmememanyskill/CYD-Klipper/actions/runs/8217091092
Branch (use to download code): https://github.com/suchmememanyskill/CYD-Klipper/tree/experimental-multi-printer
Build/Run instructions: https://github.com/suchmememanyskill/CYD-Klipper/blob/experimental-multi-printer/Build.md

**Reply by jan610** [reply]
@suchmememanyskill 
It works like a charm:
![CYD Klipper Multi Printer](https://github.com/suchmememanyskill/CYD-Klipper/assets/17181979/b6bac3aa-6430-4a9e-b57b-c023f850221c)
Once you know that you have to enable the Multiprinter feature it is verry intuitive.
It all works: adding/deleteing printer, renaming and swtiching between the controlled printer and it doesn't forget the second printer after disabeling reanabeling the Multiprinter feature again. The only two things I have for you is that i'm missing something like a title in the other tabs that tells me which printer i'm controlling currently and the settings button in the macros tab doesn't feel right now that there is another more general tab...but thats absolute nitpicking on my side. Nice job and thank you very much for this !

**[EDIT]** I just had a state where my second printer showed as offline but was Online(relay switched off) it showed as online again after i turned the relay on via browser. **[EDIT]**

**Reply by suchmememanyskill** [reply]
> The only two things I have for you is that i'm missing something like a title in the other tabs that tells me which printer i'm controlling currently

That's the idea behind being able to set the ui color per printer. I don't think there's space to constantly be displaying a name

> [EDIT] I just had a state where my second printer showed as offline but was Online(relay switched off) it showed as online again after i turned the relay on via browser. [EDIT]

Oh, yeah. If the request fails to fetch printer info the printer is considered offline. That'll take quite a bit of effort to fix :(

Thank you for the feedback!

**Reply by suchmememanyskill** [reply]
@jan610 New build

Build: https://github.com/suchmememanyskill/CYD-Klipper/actions/runs/8255107149
Instructions: https://github.com/suchmememanyskill/CYD-Klipper/blob/dev/Build.md

Printer status now says 'In Control' in the printer screen if the screen is controlling the printer.

Also, when a printer is now considered 'offline', an additional check is made against if the printer has any available power devices. If it does, the control button will be swapped for a 'Power' button that allows you to set moonraker power devices

**Reply by crazy-hair** [reply]
Absolutely brilliant! The printer list UI is super user friendly and easy to use. Haven't run into any issues, but haven't tested multiple printers printing at once yet. I'll try to get around to testing it more when I have some free time. ![video](https://i.imgur.com/ym5O2JG.gif) So far it seems to handle 5 printers just fine, we'll see how it does when they're all doing something.

**Reply by suchmememanyskill** [reply]
Just as a heads up, the build above is slightly bugged.

The macros menu's power devices will now activate power devices from other printers. Fixed this locally but haven't pushed the fix yet.

(Also sometimes buttons don't get disabled right)

**Reply by jan610** [reply]
Just tried you build from 1h ago: https://github.com/suchmememanyskill/CYD-Klipper/actions/runs/8270388222
With the power option in the printer tab we dont really need to get send back to the lost communication with MCU site, wehen turning of the relay, do we ?

**Reply by suchmememanyskill** [reply]
> With the power option in the printer tab we dont really need to get send back to the lost communication with MCU site, wehen turning of the relay, do we ?

Can you clarify what you mean by this? 

If you mean if the printer loses the connection to the printer, it'll eventually jump back to the waiting for connection screen, i don't know if i can change that. It's pretty baked into the design of CYD-Klipper (as i never expected to implement this)

Also do you think the feedback points you gave are implemented correctly?

**Reply by jan610** [reply]
Yeah i'm talikng about when the printer looses connection, it then jumps back to where we can power up the last controlled printer again or restart the firmware ect that feels a bit redundant since we have the printer tab now. when i really have a problem that needs debugging i go to back to my PC and open up the brower anyway. But again thats nitpicking and maybe i am missing a scenario where we qctually need to jump back.

I still get an the "error" state in the printer tab for the other printer that is powered down and obviously not controlled(as it's not possible). So I saw 4 states: In Controll, Idle, Offline and Error but I rarely get offline... i guess it's because my printers return "Lost communication with MCU" ?!? is that a problem on my side ?
Then when the current printer is diconnected and we jump back to that special view with the connection result output stuff we can tap on the top right home button where we can switch between printers(no clue what to call it), which is kinda redundant if we just dont jump back and just show the printer tab...There i have an offline and an Active Printer but they are both offline(power off) so does the active state overwrite the offline state cause im currently controlling it ?

Do you have questions about a specific "feedback point" ?

Again thank you verry much it's already great !

**Reply by suchmememanyskill** [reply]
> is that a problem on my side ?

That is... something i did not consider. Some people tie moonraker power devices to the klipper service (so if the power is turned off, klipper is turned off). That's what i expected you to do too. 'Offline' means that CYD-Klipper could not successfully connect to Klipper. At least currently when you try to control the error'd printer it brings you to the error ui page (where you can switch printer also if desired)

The issue from my side is that the way CYD-Klipper is designed is very much with one printer; you are 'in control' of a single printer at a time. So you see all the error state screens if you are in an error state. I'll need to redesign a whole bunch of how CYD-Klipper works to get something that's equally nice to use for single printer and multi printer operation. For now i'll just leave it as is

'Active' printer in the minimal printer list means that's the printer CYD-Klipper is currently trying to control. Its state is not displayed (as it's displayed in the 'main ui')

Either way, thank you for all the feedback :)
I'll see what i can do in the future to make this all a more coherent experience. Maybe i should just always show the navbar of buttons but just hide some depending if you can access them or not.

**Reply by suchmememanyskill** [reply]
Actually thinking about it, making the error ui into a panel and displaying it in the navbar wouldn't be too difficult. I'll give it a go tomorrow

**Reply by suchmememanyskill** [reply]
@jan610 do you mean something like this?
![PXL_20240314_185418162](https://github.com/suchmememanyskill/CYD-Klipper/assets/38142618/5c8a0871-c416-434c-9b56-67fe902aeb5a)

Build: https://github.com/suchmememanyskill/CYD-Klipper/actions/runs/8285461920


The ip connect screen is a lot harder to move so i'll leave that for a later date

**Reply by jan610** [reply]
Yeah, was talking about this:
![Printer is not ready](https://github.com/suchmememanyskill/CYD-Klipper/assets/17181979/39255c8c-2b7a-4a91-89c3-9e385bb81b8a)
I keep the Pi on and only turn of the Printer thats why it returns this.
 And i like the new build from 1h ago where we still have acess to the multi printer tab and can comfortably switch between the printers and see all the important informations i think this is already much better !

**Comment by smokeandlights** [comment]
Thanks for all the great work you've done on this project. You've implemented this really well from what I've seen. I would like to know if there's a way I can preload a BUNCH of printers in instead of adding manually. Like, can I include a table and manually flash instead of using the web installer?
